include/boost/histogram/python/register_histogram.hpp

```diff
@@ -178,82 +178,72 @@ py::class_<bh::histogram<A, S>> register_histogram(py::module &m, const char *na
             },
             "Project to a single axis or several axes on a multidiminsional histogram")
 
-        .def(make_pickle<histogram_t>())
+        .def(
+            "fill",
+            [](histogram_t &self, py::args args, py::kwargs kwargs) {
+                auto empty_weight = py::object();
+                auto weight       = optional_arg(kwargs, "weight", empty_weight);
+
+                using arrayd = py::array_t<double>;
+
+                // TODO: compute this typelist from the value types of the supported axes
+                using VArg = boost::variant2::variant<double, arrayd>;
+                auto vargs = bh::detail::make_stack_buffer<VArg>(bh::unsafe_access::axes(self));
+
+                if(args.size() != self.rank())
+                    throw std::invalid_argument("Wrong number of args");
+
+                unsigned iarg = 0;
+                for(auto arg : args) {
+                    if(py::isinstance<py::buffer>(arg) || py::hasattr(arg, "__iter__")) {
+                        auto tmp       = py::cast<arrayd>(arg);
+                        vargs.at(iarg) = tmp;
+                        if(tmp.ndim() != 1)
+                            throw std::invalid_argument("All arrays must be 1D");
+                    } else {
+                        vargs.at(iarg) = py::cast<double>(arg);
+                    }
+                    ++iarg;
+                }
 
-        ;
+                // TODO handle weights
+                using storage_t = typename histogram_t::storage_type;
+                bh::detail::static_if<boost::mp11::mp_or<std::is_same<storage_t, storage::profile>,
+                                                         std::is_same<storage_t, storage::weighted_profile>>>(
+                    [&](auto &h) {
+                        auto sample = required_arg<py::object>(kwargs, "sample");
+                        finalize_args(kwargs);
+                        auto sarray = py::cast<arrayd>(sample);
```

Comment:
We probably need to check here again that sarray has the right dimension. The size should be checked by boost::histogram.